drm-howto/modeset-atomic.c

```diff
@@ -0,0 +1,1073 @@
+			if (get_property_value(fd, props, "type") == DRM_PLANE_TYPE_PRIMARY) {
+				found_primary = true;
+				dev->plane_id = plane_id;
+				ret = 0;
+			}
+			
+			drmModeFreeObjectProperties(props);
+		}
+
+		drmModeFreePlane(plane);
+	}
+
+	drmModeFreePlaneResources(plane_res);
+
+	if (found_primary)
+		fprintf(stdout, "found primary plane, id: %d\n", dev->plane_id);
+	else
+		fprintf(stdout, "found non-primary plane, id: %d\n", dev->plane_id);
+	return ret;
+}
+
+/*
+ * modeset_create_fb() stays the same.
+ */
+
+static int modeset_create_fb(int fd, struct modeset_buf *buf)
+{
+	struct drm_mode_create_dumb creq;
+	struct drm_mode_destroy_dumb dreq;
+	struct drm_mode_map_dumb mreq;
+	int ret;
+	uint32_t handles[4] = {0}, pitches[4] = {0}, offsets[4] = {0};
+
+	/* create dumb buffer */
+	memset(&creq, 0, sizeof(creq));
+	creq.width = buf->width;
+	creq.height = buf->height;
+	creq.bpp = 32;
+	ret = drmIoctl(fd, DRM_IOCTL_MODE_CREATE_DUMB, &creq);
+	if (ret < 0) {
+		fprintf(stderr, "cannot create dumb buffer (%d): %m\n",
+			errno);
+		return -errno;
+	}
+	buf->stride = creq.pitch;
+	buf->size = creq.size;
+	buf->handle = creq.handle;
+
+	/* create framebuffer object for the dumb-buffer */
+	handles[0] = buf->handle;
+	pitches[0] = buf->stride;
+	ret = drmModeAddFB2(fd, buf->width, buf->height, DRM_FORMAT_XRGB8888,
+			    handles, pitches, offsets, &buf->fb, 0);
+	if (ret) {
+		fprintf(stderr, "cannot create framebuffer (%d): %m\n",
+			errno);
+		ret = -errno;
+		goto err_destroy;
+	}
+
+	/* prepare buffer for memory mapping */
+	memset(&mreq, 0, sizeof(mreq));
+	mreq.handle = buf->handle;
+	ret = drmIoctl(fd, DRM_IOCTL_MODE_MAP_DUMB, &mreq);
+	if (ret) {
+		fprintf(stderr, "cannot map dumb buffer (%d): %m\n",
+			errno);
+		ret = -errno;
+		goto err_fb;
+	}
+
+	/* perform actual memory mapping */
+	buf->map = mmap(0, buf->size, PROT_READ | PROT_WRITE, MAP_SHARED,
+		        fd, mreq.offset);
+	if (buf->map == MAP_FAILED) {
+		fprintf(stderr, "cannot mmap dumb buffer (%d): %m\n",
+			errno);
+		ret = -errno;
+		goto err_fb;
+	}
+
+	/* clear the framebuffer to 0 */
+	memset(buf->map, 0, buf->size);
+
+	return 0;
+
+err_fb:
+	drmModeRmFB(fd, buf->fb);
+err_destroy:
+	memset(&dreq, 0, sizeof(dreq));
+	dreq.handle = buf->handle;
+	drmIoctl(fd, DRM_IOCTL_MODE_DESTROY_DUMB, &dreq);
+	return ret;
+}
+
+/*
+ * modeset_destroy_fb() stays the same.
+ */
+
+static void modeset_destroy_fb(int fd, struct modeset_buf *buf)
+{
+	struct drm_mode_destroy_dumb dreq;
+
+	/* unmap buffer */
+	munmap(buf->map, buf->size);
+
+	/* delete framebuffer */
+	drmModeRmFB(fd, buf->fb);
+
+	/* delete dumb buffer */
+	memset(&dreq, 0, sizeof(dreq));
+	dreq.handle = buf->handle;
+	drmIoctl(fd, DRM_IOCTL_MODE_DESTROY_DUMB, &dreq);
+}
+
+/*
+ * main() also changes. Instead of performing the KMS setup calling
+ * drmModeSetCrtc(), we instead setup it using the atomic API with
+ * the function modeset_atomic_modeset().
+ */
+
+int main(int argc, char **argv)
```

Comment:
Hiding `main()` in the middle of a file is somewhat unorthodox.

The ordering of functions in a `.c` file in general should be such that static functions do not need separate declarations. This means that if function A calls function B, function B comes first in the file. It's not always possible, but at least it helps me to find things: I assume `main()`, if it exist, is found at the bottom of a file since it's the one calling everything else. If I'm reading code somewhere and want to look at the implementation of a function called from there, I expect that I should find the function somewhere above, hopefully close.

Not needing separate declarations means less code to wade through, and less places to fix if you change a function signature.

Reply:
Yeah, this ordering is a remnant of the file being an actual "tutorial", rather than something you deploy. The comments walk you through the code, hence the ordering was a bit off. Not sure this worked out as it should, though.

Reply:
I think since this example got bigger than the other, the ordering was getting in the way instead of helping. Removed the signatures and pushed the main to the bottom of the file.